machetli/search.py

```diff
@@ -86,18 +88,62 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
-                # FIXME: this is not proper error handling yet.
-                successor = None
-            if successor:
-                logging.info(successor.change_msg)
-                current_state = successor.state
-                break
+        try:
+            improving_state, message = _get_improving_successor(
+                evaluator_path, successors, environment)
+        except SubmissionError as e:
+            logging.critical(f"Terminating search because job submission for successor evaluation failed:\n{e}")
+        except PollingError as e:
+            logging.critical(f"Terminating search because querying the status of a submitted successor evaluation failed:\n{e}")
+
+        if message:
+            logging.info(message)
+        if improving_state:
+            current_state = improving_state
         else:
-            logging.info("No improving successor found, terminating search.")
             return current_state
+
+
+def _get_improving_successor(evaluator_path, successors, environment):
+    report_if_no_improvement = set()
+    def on_task_completed(job_id, task):
+        if environment.allow_nondeterministic_successor_choice:
+            # If an evaluation reports a non-improving successor,
+            # deterministic mode considers the next successor. In all
+            # other cases (improving successor, evaluation did not
+            # finish correctly), there is no need to evaluate later
+            # successors.
+            if task.status != EvaluationTask.DONE_AND_NOT_IMPROVING:
+                environment.cancel(job_id, after=task)
+        else: # non-deterministic mode
+            if task.status == EvaluationTask.DONE_AND_IMPROVING:
+                # We found an improving successor, so all other
+                # evaluations can be canceled.
+                environment.cancel(job_id)
+            elif task.status == EvaluationTask.OUT_OF_RESOURCES:
+                report_if_no_improvement.add(task)
+            elif (task.status == EvaluationTask.CRITICAL):
+                logging.warn(f"Critical error in task {job_id}_{task.task_id}: {task.error}")
+
+    for batch in batched(successors, environment.batch_size):
+        tasks = environment.run(evaluator_path, batch, on_task_completed)
+        for task in tasks:
+            if task.status == EvaluationTask.DONE_AND_NOT_IMPROVING:
+                continue
+            elif task.status == EvaluationTask.DONE_AND_IMPROVING:
+                return task.successor.state, task.successor.change_msg
+            elif not environment.allow_nondeterministic_successor_choice:
+                return None, (
+                    "Errors occurred in successor evaluations before finding "
+                    "an improving successor. With the option "
+                    "`allow_nondeterministic_successor_choice` an improving "
+                    "successor found later would not count.")
```

Comment:
Should we report here what kind of error occurred? Or do we do so somewhere else? Also, the error message confuses me: With the option, we *would* consider a successor found later, wouldn't we?